### PR TITLE
the EFI changes from PR #9437 (the boot counting PR)

### PR DESCRIPTION
**Pull request by poettering**
As requested by @keszybz, just the sd-boot EFI code related commits from PR #9437

### CODE REVIEW

File: src/boot/efi/util.h [hunk 1]
```diff
@@ -5,6 +5,7 @@
 #include <efilib.h>
 
 #define ELEMENTSOF(x) (sizeof(x)/sizeof((x)[0]))
+#define OFFSETOF(x,y) ((UINTN) &(((x*)0)->y))
```
Comment (keszybz) [inline src/boot/efi/util.h:8]: Hmm, this might be undefined behaviour: https://software.intel.com/en-us/blogs/2015/04/20/null-pointer-dereferencing-causes-undefined-behavior, "About C". Why not define it as in /usr/lib/gcc/x86_64-redhat-linux/8/include/stddef.h:
```C
#define offsetof(TYPE, MEMBER) __builtin_offsetof (TYPE, MEMBER)
```
?

File: src/boot/efi/boot.c [hunk 2]
```diff
@@ -1245,12 +1457,44 @@ static VOID config_load_entries(
...
 }
 
+static INTN config_entry_compare(ConfigEntry *a, ConfigEntry *b) {
+        INTN r;
+
+        /* Order entries that have no tries left to the end of the list */
+        if (a->tries_left != 0 && b->tries_left == 0)
+                return -1;
+        if (a->tries_left == 0 && b->tries_left != 0)
+                return 1;
```
Comment (keszybz) [inline src/boot/efi/boot.c:1473]: Ah, I was worried what happens if we boot a kernel but the "blessing mechanism" which is not part of this patch is inactive. The counter goes to 0, but we don't completely skip this entry, just make it lower priority.

File: src/boot/efi/util.h (hunk 1 above)
Comment (poettering) [reply src/boot/efi/util.h:8]: Indeed! Posted #10451 that adds this
